### PR TITLE
Pdb 25 account creation (student flow)

### PR DESCRIPTION
**Pull request by heidage**
1) Added in JWT passing between frontend and backend
2) Added in account initialisation after clicking either Student / Tutor
3) Added in account creation depending on which account type user clicks on

### CODE REVIEW

File: Bamboo/frontend/bamboo-frontend/app/routes/_setup+/welcome.tsx [hunk 1]
```diff
@@ -30,6 +54,11 @@ export default function SetupWelcomePage(){
   const handleButtonClick = () => {
     if(selectState){
       console.log(`SUBMIT RESPONSE HERE!, ${selectState}`)
+      auth.initilaise.accountType(
+        BACKEND,
+        selectState,
+        jwt
```
Comment (TsienJin) [inline Bamboo/frontend/bamboo-frontend/app/routes/_setup+/welcome.tsx:60]: iirc http-only cookies will be sent to the backend if we set "auth:true" when making a request

File: Bamboo/backend/bamboo-backend/src/routes/auth/initialise/initialise.controller.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+    @Post()
+    @UseInterceptors(FormDataInterceptor)
+    async initialiseAccountType(
+        @Body() body,
+        @Res() res
+    ){
+        const decoded: JwtPayload = jwt.decode(body.jwt) as JwtPayload;
+        try {
+            const { accountJwt } = await this.initialiseService.initialiseAccount(
```
Comment (TsienJin) [inline Bamboo/backend/bamboo-backend/src/routes/auth/initialise/initialise.controller.ts:27]: I'll combine the JWT logic to a middleware or interceptor so this code can be isolated

File: Bamboo/backend/bamboo-backend/src/routes/auth/initialise/initialise.service.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+    async initialiseAccount(accountType: string, jwt: JwtPayload){
+        const account = await this.usersService.initialiseAccount(
+            accountType,
+            jwt
+        )
+
+        // Generate a new JWT token with the updated account
+        const firstName = account.name.split(' ')[0]
+        const lastName = account.name.split(' ')[1]
```
Comment (TsienJin) [inline Bamboo/backend/bamboo-backend/src/routes/auth/initialise/initialise.service.ts]: dont use string.split methods when determining firstName and lastName. These should be separate attributes

Comment (TsienJin) [reply Bamboo/backend/bamboo-backend/src/routes/auth/initialise/initialise.service.ts]: userService.initialiseAccount response type should match the response type of userService.createUser or .findOrCreate

File: Bamboo/backend/bamboo-backend/src/services/user/user.service.ts [hunk 4]
```diff
@@ -49,4 +50,41 @@ export class UserService {
...
+          email: jwt.email
+        },
+        data:{
+          initialised: true
+        }
+      }
+    )
+
+    let account = null
+    if (accountType === 'student') {
```
Comment (TsienJin) [inline Bamboo/backend/bamboo-backend/src/services/user/user.service.ts]: Use switch instead

File: Bamboo/backend/bamboo-backend/src/services/user/user.service.ts
Comment (TsienJin) [inline Bamboo/backend/bamboo-backend/src/services/user/user.service.ts:1]: Reference the code on the latest staging -- do `npm run prisma` if u've yet to run that. The updated schema should allow for optional passwords. Setting password as `''`  will result in a hash for an empty string. This is a vulnerability as a user can use an empty string to log in.

File: Bamboo/backend/bamboo-backend/src/services/user/user.service.ts [hunk 5]
```diff
@@ -49,4 +50,41 @@ export class UserService {
...
+          email: jwt.email,
+          password: '',
+        }
+      })
+    }
+
+    if (accountType === 'tutor') {
+      account = await this.prismaService.prisma.tutor.create({
+        data: {
+          name: `${jwt.firstName} ${jwt.lastName}`,
```
Comment (TsienJin) [inline Bamboo/backend/bamboo-backend/src/services/user/user.service.ts]: use firstName and lastName instead

File: Bamboo/backend/bamboo-backend/src/services/user/user.service.ts [hunk 6]
```diff
@@ -49,4 +50,41 @@ export class UserService {
     return existingUser
   
   }
+
+  async initialiseAccount(accountType: string, jwt: JwtPayload){
```
Comment (TsienJin) [inline Bamboo/backend/bamboo-backend/src/services/user/user.service.ts]: initialiseAccount should not contain logic to create the `Student`/`Tutor` record in the DB -- this should be done in a separate service; one for student (e.g. studentService) and one for tutor (e.g. tutorService)

File: Bamboo/backend/bamboo-backend/src/services/user/user.service.ts
Comment (heidage) [reply Bamboo/backend/bamboo-backend/src/services/user/user.service.ts:1]: I was thinking of using the code from the lastest staging but the schema.prisma doesnt have `password String?` for `Tutor` and `Student`.The `name` as well is not `firstName` and `lastName`. I was thinking maybe we could use the User object for registration and login, and having the emil as secondary key hence using it which links us to individual `Student` or `Tutor` record.

File: Bamboo/frontend/bamboo-frontend/app/routes/_setup+/welcome.tsx (hunk 1 above)
Comment (TsienJin) [reply Bamboo/frontend/bamboo-frontend/app/routes/_setup+/welcome.tsx:60]: Set to be HTTP only "false" for debugging

**Pull request merged**